File: Components/Document Server/VB/DocsMainDemo/Modules/RichEditMailMerge.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.IO
Imports System.Data
Imports DevExpress.XtraRichEdit
Imports System.Collections.Generic
Imports DevExpress.Office.Services
Imports DevExpress.DXperience.Demos
Imports DevExpress.XtraPrintingLinks

Namespace DevExpress.Docs.Demos
	Partial Public Class RichEditMailMerge
		Inherits TutorialControlBase
		Private link As PrintableComponentLinkBase
		Private documentServer As RichEditDocumentServer
		Private connection As New System.Data.OleDb.OleDbConnection()
		Public Sub New()
			InitializeComponent()

			Me.documentServer = New RichEditDocumentServer()
			Dim path As String = DemoUtils.GetRelativePath("MailMerge.rtf")
			Me.documentServer.LoadDocument(path, DocumentFormat.Rtf)

			Me.printPreviewControl.PrintingSystem = New DevExpress.XtraPrinting.PrintingSystem()
			Me.link = New DevExpress.XtraPrintingLinks.PrintableComponentLinkBase(Me.printPreviewControl.PrintingSystem)
			LoadData()
			Me.documentServer.Options.MailMerge.ViewMergedData = True
			AddHandler Me.gridView1.FocusedRowChanged, AddressOf gridView1_FocusedRowChanged
			AddHandler Load, AddressOf MailMerge_Load
		End Sub

		Private Sub MailMerge_Load(ByVal sender As Object, ByVal e As EventArgs)
			ProcessSelectedRows()
		End Sub
		Private Sub ProcessSelectedRows()
			UpdateMailMergeOptions()
			link.Component = Me.documentServer
			link.CreateDocument()
		End Sub
		Protected Sub UpdateMailMergeOptions()
			Dim selectedRowIndexes() As Integer = gridView1.GetSelectedRows()
			Dim count As Integer = selectedRowIndexes.Length
			Dim selectedRows As New List(Of Object)()
			For i As Integer = 0 To count - 1
				selectedRows.Add(gridView1.GetRow(selectedRowIndexes(i)))
			Next i
			Me.documentServer.Options.MailMerge.DataSource = selectedRows
		End Sub
		Private Sub gridView1_FocusedRowChanged(ByVal sender As Object, ByVal e As XtraGrid.Views.Base.FocusedRowChangedEventArgs)
			ProcessSelectedRows()
		End Sub
		Private Sub LoadData()
			Dim path As String = DemoUtils.GetRelativePath("nwind.mdb")
			If path.Length <= 0 Then
				Return
			End If
			Dim uriService As IUriStreamService = CType(Me.documentServer.GetService(GetType(IUriStreamService)), IUriStreamService)
			uriService.RegisterProvider(New DBUriStreamProvider(nwindDataSet.Employees, "Photo"))
			employeesTableAdapter.Connection = New System.Data.OleDb.OleDbConnection()
			DemoUtils.SetConnectionString(employeesTableAdapter.Connection, path)
			DemoUtils.SetConnectionString(Me.connection, path)
			Me.employeesTableAdapter.Fill(Me.nwindDataSet.Employees)
			FillData()
		End Sub
		Private Sub FillData()
			Dim cmd As String = "SELECT Employees.*, Customers.* FROM (Employees INNER JOIN EmployeeCustomers ON Employees.EmployeeID = EmployeeCustomers.EmployeeId) INNER JOIN Customers ON EmployeeCustomers.CustomerId = Customers.CustomerID;"
			Dim selectCommand As New System.Data.OleDb.OleDbCommand(cmd, connection)
			Dim adapter As New System.Data.OleDb.OleDbDataAdapter(selectCommand)
			Dim dataSet As New DataSet()
			adapter.Fill(dataSet)
			Me.gridControl1.DataSource = dataSet.Tables(0)
			Me.documentServer.Options.MailMerge.DataSource = dataSet.Tables(0)
		End Sub
	End Class
	Public Class DBUriStreamProvider
		Implements IUriStreamProvider
		Private Shared ReadOnly prefix As String = "dbimg://"
		Private table As DataTable
		Private columnName As String

		Public Sub New(ByVal table As DataTable, ByVal columnName As String)
			Me.table = table
			Me.columnName = columnName
		End Sub

		#Region "IUriStreamProvider Members"
		Private Function GetStream(ByVal uri As String) As Stream Implements IUriStreamProvider.GetStream
			uri = uri.Trim()
			If (Not uri.StartsWith(prefix)) Then
				Return Nothing
			End If
			Dim strId As String = uri.Substring(prefix.Length).Trim()
			Dim id As Integer
			If (Not Integer.TryParse(strId, id)) Then
				Return Nothing
			End If
			Dim row As DataRow = table.Rows.Find(id)
			If row Is Nothing Then
				Return Nothing
			End If
			Dim bytes() As Byte = TryCast(row(columnName), Byte())
			If bytes Is Nothing Then
				Return Nothing
			End If
			Return New MemoryStream(bytes)
		End Function
		#End Region
	End Class
End Namespace

File: Components/Document Server/CS/DocsMainDemo/Modules/RichEditMailMerge.cs
using System;
using System.IO;
using System.Data;
using DevExpress.XtraRichEdit;
using System.Collections.Generic;
using DevExpress.Office.Services;
using DevExpress.DXperience.Demos;
using DevExpress.XtraPrintingLinks;

namespace DevExpress.Docs.Demos {
    public partial class RichEditMailMerge : TutorialControlBase {
        PrintableComponentLinkBase link;
        RichEditDocumentServer documentServer;
        System.Data.OleDb.OleDbConnection connection = new System.Data.OleDb.OleDbConnection();
        public RichEditMailMerge() {
            InitializeComponent();
            
            this.documentServer = new RichEditDocumentServer();
            string path = DemoUtils.GetRelativePath("MailMerge.rtf");
            this.documentServer.LoadDocument(path, DocumentFormat.Rtf);

            this.printPreviewControl.PrintingSystem = new DevExpress.XtraPrinting.PrintingSystem();
            this.link = new DevExpress.XtraPrintingLinks.PrintableComponentLinkBase(this.printPreviewControl.PrintingSystem);
            LoadData();
            this.documentServer.Options.MailMerge.ViewMergedData = true;
            this.gridView1.FocusedRowChanged += gridView1_FocusedRowChanged;
            Load += MailMerge_Load;
        }

        void MailMerge_Load(object sender, EventArgs e) {
            ProcessSelectedRows();
        }
        void ProcessSelectedRows() {
            UpdateMailMergeOptions();
            link.Component = this.documentServer;
            link.CreateDocument();
        }
        protected void UpdateMailMergeOptions() {
            int[] selectedRowIndexes = gridView1.GetSelectedRows();
            int count = selectedRowIndexes.Length;
            List<object> selectedRows = new List<object>();
            for (int i = 0; i < count; i++)
                selectedRows.Add(gridView1.GetRow(selectedRowIndexes[i]));
            this.documentServer.Options.MailMerge.DataSource = selectedRows;
        }
        void gridView1_FocusedRowChanged(object sender, XtraGrid.Views.Base.FocusedRowChangedEventArgs e) {
            ProcessSelectedRows();
        }
        void LoadData() {
            string path = DemoUtils.GetRelativePath("nwind.mdb");
            if (path.Length <= 0)
                return;
            IUriStreamService uriService = (IUriStreamService)this.documentServer.GetService(typeof(IUriStreamService));
            uriService.RegisterProvider(new DBUriStreamProvider(nwindDataSet.Employees, "Photo"));
            employeesTableAdapter.Connection = new System.Data.OleDb.OleDbConnection();
            DemoUtils.SetConnectionString(employeesTableAdapter.Connection, path);
            DemoUtils.SetConnectionString(this.connection, path);
            this.employeesTableAdapter.Fill(this.nwindDataSet.Employees);
            FillData();
        }
        void FillData() {
            string cmd = @"SELECT Employees.*, Customers.* FROM (Employees INNER JOIN EmployeeCustomers ON Employees.EmployeeID = EmployeeCustomers.EmployeeId) INNER JOIN Customers ON EmployeeCustomers.CustomerId = Customers.CustomerID;";
            System.Data.OleDb.OleDbCommand selectCommand = new System.Data.OleDb.OleDbCommand(cmd, connection);
            System.Data.OleDb.OleDbDataAdapter adapter = new System.Data.OleDb.OleDbDataAdapter(selectCommand);
            DataSet dataSet = new DataSet();
            adapter.Fill(dataSet);
            this.gridControl1.DataSource = dataSet.Tables[0];
            this.documentServer.Options.MailMerge.DataSource = dataSet.Tables[0];
        }
    }
    public class DBUriStreamProvider : IUriStreamProvider {
        static readonly string prefix = "dbimg://";
        DataTable table;
        string columnName;

        public DBUriStreamProvider(DataTable table, string columnName) {
            this.table = table;
            this.columnName = columnName;
        }

        #region IUriStreamProvider Members
        Stream IUriStreamProvider.GetStream(string uri) {
            uri = uri.Trim();
            if (!uri.StartsWith(prefix))
                return null;
            string strId = uri.Substring(prefix.Length).Trim();
            int id;
            if (!int.TryParse(strId, out id))
                return null;
            DataRow row = table.Rows.Find(id);
            if (row == null)
                return null;
            byte[] bytes = row[columnName] as byte[];
            if (bytes == null)
                return null;
            return new MemoryStream(bytes);
        }
        #endregion
    }
}
